STM32F4xx modifications for HAL_STM32

ktand opened the pull request:
Since the addition of the **HAL_STM32** HAL for **Arduino Core STM32** the **HAL_STM32F4** no longer builds with that core. This PR contains modifications to make the STM32F4 work with the **HAL_STM32** instead.

This includes:

- Support for external I2C/SPI eeprom
- Optional configuration of timer selection and priorities in pins_XXX.h
- Use timer defaults for STM32F4xx from HAL_STM32F4.
- Exclusion of HAL_STM32F4 files when building for Arduino Core STM32.
- STM32F4xx now uses `HAL_TIMER_RATE` of `F_CPU/2` instead of `HAL_RCC_GetPCLK2Freq()` because it's much faster. `F_CPU` is updated by the core if/when frequency changes.

ktand committed:
Support SPI & I2C eeprom

ktand committed:
Exclude HAL_STM32F4 files when building for Arduino Core STM32

ktand committed:
STM32F4 fixes for HAL_STM32
Removed some unused & duplicated defines

thinkyhead commented:
CC: @xC0000005 @p3p @hasenbanck @alexxy

hasenbanck commented on Marlin/src/HAL/HAL_STM32/HAL_timers_STM32.h:
```diff
@@ -34,42 +34,66 @@
 #elif defined STM32F4xx
 
-  #define HAL_TIMER_RATE (HAL_RCC_GetPCLK2Freq()) // frequency of timer peripherals
+  #define HAL_TIMER_RATE (F_CPU/2) // frequency of timer peripherals
```
Now the way we read the frequency of the CPU by two different ways: F_CPU and HAL_RCC_GetSysClockFreq() (F1, F4, F7). I think we should only use one or the other. I'm fine with either.

ktand replied:
Actually there are three different ways. F0 and F7 uses  `HAL_RCC_GetSysClockFreq()`, F1 uses `HAL_RCC_GetPCLK2Freq()` and my change `F_CPU`. 

F_CPU is just a define for a global varable, SystemCoreClock, that is updated by the core when/if the clock changes. 

Calling  HAL_RCC_GetSysClockFreq() is time consuming (especially as this is called from the stepper isr, when getting the timer counter for the stepper pulse timing):

```
__weak uint32_t HAL_RCC_GetSysClockFreq(void)
{
  uint32_t pllm = 0U, pllvco = 0U, pllp = 0U;
  uint32_t sysclockfreq = 0U;

  /* Get SYSCLK source -------------------------------------------------------*/
  switch (RCC->CFGR & RCC_CFGR_SWS)
  {
    case RCC_CFGR_SWS_HSI:  /* HSI used as system clock source */
    {
      sysclockfreq = HSI_VALUE;
       break;
    }
    case RCC_CFGR_SWS_HSE:  /* HSE used as system clock  source */
    {
      sysclockfreq = HSE_VALUE;
      break;
    }
    case RCC_CFGR_SWS_PLL:  /* PLL used as system clock  source */
    {
      /* PLL_VCO = (HSE_VALUE or HSI_VALUE / PLLM) * PLLN
      SYSCLK = PLL_VCO / PLLP */
      pllm = RCC->PLLCFGR & RCC_PLLCFGR_PLLM;
      if(__HAL_RCC_GET_PLL_OSCSOURCE() != RCC_PLLSOURCE_HSI)
      {
        /* HSE used as PLL clock source */
        pllvco = ((HSE_VALUE / pllm) * ((RCC->PLLCFGR & RCC_PLLCFGR_PLLN) >> POSITION_VAL(RCC_PLLCFGR_PLLN)));
      }
      else
      {
        /* HSI used as PLL clock source */
        pllvco = ((HSI_VALUE / pllm) * ((RCC->PLLCFGR & RCC_PLLCFGR_PLLN) >> POSITION_VAL(RCC_PLLCFGR_PLLN)));    
      }
      pllp = ((((RCC->PLLCFGR & RCC_PLLCFGR_PLLP) >> POSITION_VAL(RCC_PLLCFGR_PLLP)) + 1U) *2U);
      
      sysclockfreq = pllvco/pllp;
      break;
    }
    default:
    {
      sysclockfreq = HSI_VALUE;
      break;
    }
  }
  return sysclockfreq;
}
```
I guess there could be issues of using F_CPU depending on clock config, but then I guess the same applies to `HAL_RCC_GetSysClockFreq`. 

What do you suggest?

hasenbanck replied:
I don't think saving some CPU cycles in this part of the code is that important, since the calculated values are only used for setting up the timers. They are not used in an interrupt or the main loop as far as I know.

The current assumption we take for the timer clocks are good educated guesses I think. If a board designer decided to use a weird timer clock, then he/she can still change that in the board definition.

ktand replied:
It is used in the stepper irq:

```cpp
hal_timer_t pulse_end = HAL_timer_get_count(PULSE_TIMER_NUM) + hal_timer_t(MIN_PULSE_TICKS);
const hal_timer_t added_step_ticks = hal_timer_t(ADDED_STEP_TICKS);
```

`MIN_PULSE_TICKS` and `ADDED_STEP_TICKS` uses `PULSE_TIMER_PRESCALE` which in turn uses `HAL_TIMER_RATE`.

hasenbanck commented on Marlin/src/HAL/HAL_STM32/watchdog_STM32.cpp:
```diff
@@ -31,7 +31,14 @@
+    IWatchdog.reload(); 
+
+    #if PIN_EXISTS(LED)
+      TOGGLE(LED_PIN);  // heart beat indicator
```
Is this leftover debug code or do you want to really change the standard behavior of the LED?

ktand replied:
For **HAL_LPC1767**, **HAL_STM32F1** and **HAL_STM32F4** there is a "heart beat" indicator using the LED pin. I don't think the LED is used for anything else except for endstop debugging when `PINS_DEBUGGING `is enabled, but correct me if I'm wrong. I think it's very nice to see that the board is alive.

I could change the implementation like this:

```
    #if !defined(PINS_DEBUGGING && PIN_EXISTS(LED)
      TOGGLE(LED_PIN);  // heart beat indicator
    #endif
```

Or, what about an implementation using a macro, like the `HAL_timer_isr_prologue/HAL_timer_isr_epilogue`? E.g "`Watchdog_reset_epilog`" ?

```
  void watchdog_reset() 
  {
    IWatchdog.reload(); 

   Watchdog_reset_epilog();
  }
```

hasenbanck replied:
I think that we should at least catch the case with the pin debug. 

So
```
    #if !defined(PINS_DEBUGGING) && PIN_EXISTS(LED)
      TOGGLE(LED_PIN);  // heart beat indicator
    #endif
```

Seems like a good idea. You did miss a closing bracket, which I added in my copy.

hasenbanck commented on Marlin/src/pins/pins_REMRAM_V1.h:
```diff
@@ -124,3 +124,10 @@
+//
+
+#define STEP_TIMER 2
+#define TEMP_TIMER_IRQ_PRIO 6
```
The board doesn't need an exception for the IRQ PRIO. I think I did change the priority at default when I changed the timer code from @xC0000005 . So the standard value is fine.

ktand replied:
Ok, I'll remove this.

xC0000005 commented on Marlin/src/HAL/HAL_STM32/HAL_timers_STM32.cpp:
```diff
@@ -79,15 +79,15 @@ void HAL_timer_start(const uint8_t timer_num, const uint32_t frequency) {
         TimerHandle[timer_num].irqHandle = Step_Handler;
         TimerHandleInit(&TimerHandle[timer_num], (((HAL_TIMER_RATE) / step_prescaler) / frequency) - 1, step_prescaler);
-        HAL_NVIC_SetPriority(STEP_TIMER_IRQ_NAME, 6, 0);
+        HAL_NVIC_SetPriority(STEP_TIMER_IRQ_NAME, STEP_TIMER_IRQ_PRIO, 0);
```
This is a reasonable change, and the right thing to do, so long as we define a sane default.

xC0000005 commented on Marlin/src/HAL/HAL_STM32F4/HAL_timers_STM32F4.cpp:
```diff
@@ -91,7 +91,7 @@ void HAL_timer_start(const uint8_t timer_num, const uint32_t frequency) {
           TimerHandleInit(&TimerHandle[timer_num], (((HAL_TIMER_RATE) / step_prescaler) / frequency) - 1, step_prescaler);
         #endif
-        HAL_NVIC_SetPriority(STEP_TIMER_IRQ_ID, 6, 0);
+        HAL_NVIC_SetPriority(STEP_TIMER_IRQ_ID, 1, 0);
```
Why is one priority set with a per board priority/#define and this one is hard coded? I'm not saying "Don't," I'm just trying to understand why.

ktand commented:
*@xC0000005, that is a change to the STM32F4 HAL, not the generic one.*

Regards
Karl
On Sun, Oct 14, 2018 at 7:37 AM J.C. Nelson <notifications@github.com>
wrote:

> *@xC0000005* commented on this pull request.
> ------------------------------
>
> In Marlin/src/HAL/HAL_STM32F4/HAL_timers_STM32F4.cpp
> <https://github.com/MarlinFirmware/Marlin/pull/12080#discussion_r224980162>
> :
>
> > @@ -91,7 +91,7 @@ void HAL_timer_start(const uint8_t timer_num, const uint32_t frequency) {
>            TimerHandle[timer_num].irqHandle = TC5_Handler;
>            TimerHandleInit(&TimerHandle[timer_num], (((HAL_TIMER_RATE) / step_prescaler) / frequency) - 1, step_prescaler);
>          #endif
> -        HAL_NVIC_SetPriority(STEP_TIMER_IRQ_ID, 6, 0);
> +        HAL_NVIC_SetPriority(STEP_TIMER_IRQ_ID, 1, 0);
>
> Why is one priority set with a per board priority/#define and this one is
> hard coded? I'm not saying "Don't," I'm just trying to understand why.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/12080#pullrequestreview-164485425>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQ8NJJO9el1Igv5z2l3GhFiI37V7Sf6Aks5uks2pgaJpZM4Xafwu>
> .
>

xC0000005 commented:
Ok. Sorry, I missed that.

> On Oct 13, 2018, at 10:53 PM, Karl Andersson <notifications@github.com> wrote:
> 
> *@xC0000005, that is a change to the STM32F4 HAL, not the generic one.*
> 
> Regards
> Karl
> On Sun, Oct 14, 2018 at 7:37 AM J.C. Nelson <notifications@github.com>
> wrote:
> 
> > *@xC0000005* commented on this pull request.
> > ------------------------------
> >
> > In Marlin/src/HAL/HAL_STM32F4/HAL_timers_STM32F4.cpp
> > <https://github.com/MarlinFirmware/Marlin/pull/12080#discussion_r224980162>
> > :
> >
> > > @@ -91,7 +91,7 @@ void HAL_timer_start(const uint8_t timer_num, const uint32_t frequency) {
> > TimerHandle[timer_num].irqHandle = TC5_Handler;
> > TimerHandleInit(&TimerHandle[timer_num], (((HAL_TIMER_RATE) / step_prescaler) / frequency) - 1, step_prescaler);
> > #endif
> > - HAL_NVIC_SetPriority(STEP_TIMER_IRQ_ID, 6, 0);
> > + HAL_NVIC_SetPriority(STEP_TIMER_IRQ_ID, 1, 0);
> >
> > Why is one priority set with a per board priority/#define and this one is
> > hard coded? I'm not saying "Don't," I'm just trying to understand why.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/MarlinFirmware/Marlin/pull/12080#pullrequestreview-164485425>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AQ8NJJO9el1Igv5z2l3GhFiI37V7Sf6Aks5uks2pgaJpZM4Xafwu>
> > .
> >
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/MarlinFirmware/Marlin/pull/12080#issuecomment-429598878>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AeppccorA117dohW8_NMBoH6RpDAEAUFks5uktFTgaJpZM4Xafwu>.
>

ffnull commented:
> Since the addition of the **HAL_STM32** HAL for **Arduino Core STM32** the **HAL_STM32F4** no longer builds with that core. This PR contains modifications to make the STM32F4 work with the **HAL_STM32** instead.
> 
> This includes:
> 
> * Support for external I2C/SPI eeprom
> * Optional configuration of timer selection and priorities in pins_XXX.h
> * Use timer defaults for STM32F4xx from HAL_STM32F4.
> * Exclusion of HAL_STM32F4 files when building for Arduino Core STM32.
> * STM32F4xx now uses `HAL_TIMER_RATE` of `F_CPU/2` instead of `HAL_RCC_GetPCLK2Freq()` because it's much faster. `F_CPU` is updated by the core if/when frequency changes.

Can you help me with compiling for steval-3dp001v1 ????

ktand commented:
@thinkyhead @xC0000005 @p3p @hasenbanck @alexxy

If there are no more comments I'd be happy if this could be merged.

hasenbanck commented:
Awww, my comments haven't been posted by the system yet. I have entered
them two days ago but it seems I need to close my code review to do so. I'm
currently on the go, so I will be able to post them in roughly two hours.

>

xC0000005 commented:
The reason I originally had the macros like this was that I looked at how example code from ST set timer values, and this derives it from the actual executing config as opposed to the board’s defined limits (I am coding for a board that runs at 48mhz rather than the 72 it’s capable of, sadly). Perhaps a middle ground would be to read this once at startup and use the read value?

> On Oct 15, 2018, at 10:45 AM, Karl Andersson <notifications@github.com> wrote:
> 
> @ktand commented on this pull request.
> 
> In Marlin/src/HAL/HAL_STM32/HAL_timers_STM32.h <https://github.com/MarlinFirmware/Marlin/pull/12080#discussion_r225256711>:
> 
> >  
>  #elif defined STM32F4xx
>  
> -  #define HAL_TIMER_RATE (HAL_RCC_GetPCLK2Freq()) // frequency of timer peripherals
> +  #define HAL_TIMER_RATE (F_CPU/2) // frequency of timer peripherals
> Actually there are three different ways. STM32F0xx uses HAL_RCC_GetSysClockFreq()....
> 
> I think the F_CPU/2 method is better than calling HAL_RCC_GetPCLK2Freq() because it saves quite some CPU cycles. F_CPU will just read a global variable defined in the STM core. BUT, depending on clock configuration I think it may be that F_CPU/2 not always equals HAL_RCC_GetPLCK2Freq() so it might be best to leave the code as is, e.g use HAL_RCC_GetPCLK2Freq(). I'll revert this change.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/MarlinFirmware/Marlin/pull/12080#discussion_r225256711>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AeppcTrmU8hFFX6kBMto_SdaITJehwNVks5ulMmlgaJpZM4Xafwu>.
>

hasenbanck commented:
> The reason I originally had the macros like this was that I looked at how example code from ST set timer values, and this derives it from the actual executing config as opposed to the board’s defined limits (I am coding for a board that runs at 48mhz rather than the 72 it’s capable of, sadly). Perhaps a middle ground would be to read this once at startup and use the read value?

I'm pretty sure that it's only used in the setup methods.

hasenbanck commented:
Yeah, just looked through the code. "HAL_TIMER_RATE" is only used by the HALs to setup the timers. Some HALs use it to define "TEMP_TIMER_RATE" and "STEPPER_TIMER_RATE" which are used in some places in the main loop, but since we are using hard coded values (72khz and 2Mhz) we don't have to worry about that.

hasenbanck commented:
Damn, didn't saw that PULSE_TIMER_PRESCALE uses the HAL_TIMER_RATE value. That
unfortunate.

ktand commented:
I've had issue where the step pulse length was incorrect when using the function to get the clock because the timer count is fetched before the X_STEP is performed. I think it might be caused by

`const hal_timer_t added_step_ticks = hal_timer_t(ADDED_STEP_TICKS);`

is called AFTER getting the count which takes some cycles to complete if HAL_TIMER_RATE is defined using a function.

I would suggest changing HAL_TIMER_RATE to use F_CPU. I can confirm that F_CPU/2 is correct for the F4 when using the STM Core default clock config. Do you know what divisor to use for F0, F1 and F7?

hasenbanck commented:
@ktand I fired up STM32CubeMX and default config:
For F7 it's normally FCPU/2
F0 and F1 normally use FCPU

ktand committed:
Don't toggle heart beat led if PINS_DEBUGGING is defined
Use F_CPU instead of calling core functions to get timer clock
Remove changes to pins_REMRAM_V1.h

ktand commented:
Thanks @hasenbanck . 

I've made the following changes:

- Don't toggle heart beat led if PINS_DEBUGGING is defined
- Use F_CPU instead of calling core functions to get timer clock
- Remove changes to pins_REMRAM_V1.h

What do you think?

hasenbanck commented:
Please don't patch out the
#define STEP_TIMER 2
In the RemRam pin definition, since that one is needed. Only the last
define is not needed.

The rest look good. Thank you for taking the feedback into consideration.

ktand committed:
REMRAM uses Step Timer 2

ktand commented:
@hasenbanck Sorry about this. Fixed now.

@thinkyhead I think we're ready to have this merged now.

ffnull commented:
@ktand so can you help?

ktand commented:
@ffnul,l to build using the official STM32 core there has to be a variant
definition for the board. Is there such a definition for the
steval-3dp001v1?


On Thu, Oct 18, 2018 at 9:46 PM ffnull <notifications@github.com> wrote:

> @ktand <https://github.com/ktand> so can you help?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/12080#issuecomment-431136430>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQ8NJLw8YYlern4STCtWUJ02G5fpTU6Sks5umNqLgaJpZM4Xafwu>
> .
>

ffnull commented:
@ktand nope 
Here is pdf https://www.st.com/content/ccc/resource/technical/layouts_and_diagrams/schematic_pack/group0/a4/27/eb/f1/4d/1d/4f/c6/STEVAL-3DP001V1%20schematic/files/steval-3dp001v1_schematic.pdf/jcr:content/translations/en.steval-3dp001v1_schematic.pdf

hasenbanck commented:
@ffnull That's not a variant definition for the official STM32 core. 

Please have a look here:
https://github.com/stm32duino/Arduino_Core_STM32/issues/347
